### DISCUSSION TITLE
Does Editor have TabMode similar to Tampermonkey?

### DISCUSSION THREAD

**Discussion by Rikk** [violentmonkey/violentmonkey#1827] [Q&A] [answered]
Hello, does the editor of Violentmonkey have a 'TabMode' similar that of Tampermonkey?
There, 'TabMode: Indent' automagically indents selected code when Tab key is pressed.
I'm planning to move to Violentmonkey.

**Comment by tophf** [comment]
Yes, it's a standard built-in feature of CodeMirror, the editor used both by Tampermonkey and Violentmonkey.

**Reply by Rikk** [reply]
How to enable it?

**Reply by tophf** [reply]
It's enabled by default.

**Reply by Rikk** [reply]
Perhaps we are talking about different things... When this option is enabled in TM (TabMode: Indent), when you select a portion of code and press Tab, it indents that selection automatically, but if you press Tab more times nothing happens because code is already indented properly. My VM behaves like a normal editor, ie, pressing Tab repeatedly keeps adding indentation. (same for shift+tab)

**Reply by tophf** [reply]
Violentmonkey behaves the way you described by default. There's no setting to disable it, and I don't see why TM offers it. We respect the author's intent to indent like most (all?) editors do.

**Reply by Rikk** [reply]
Actually, I like that tab mode of TM, because it is very practical.

> I don't see why TM offers it

It provides 3 different behaviors (TabMode: Classic/Smart/Indent). I like the Indent one (that's the one we're talking about).

I've walked around in TM code, I think they use custom functions to control these behaviors of CodeMirror editor, although I couldn't understand the mechanism. I think these are the related codes:
https://github.com/Tampermonkey/tampermonkey/blob/master/src/options.js#L1429
https://github.com/Tampermonkey/tampermonkey/blob/master/src/cm/codemirror.js#L184
https://github.com/Tampermonkey/tampermonkey/blob/master/src/cm/codemirror.js#L1446

From what I understand, Violentmonkey uses CodeMirror defaults, right?

**Reply by tophf** [reply, marked as answer]
Okay, I've just tried Tampermonkey and two of its three modes for TabMode don't make sense to me because none of the popular editors works like that, AFAIK, so we definitely won't implement this option.

Violentmonkey implements the standard behavior found in popular editors when there is a selection: Tab key shifts it whole to the right (i.e. uniform indent to the right), Shift-Tab shifts the whole thing to the left.

The functionality you described seems a subset of what is usually called Reformat or Beautify and is invoked separately, not via the Tab key. VM doesn't have it and we don't plan to add it to our built-in editor, but someday we plan to make it possible to seamlessly integrate an external editor like VSCode online. Currently we only offer [semi-integration with local editors](https://violentmonkey.github.io/posts/how-to-edit-scripts-with-your-favorite-editor/).
